### PR TITLE
feat(*): extended VM support

### PR DESCRIPTION
**Pull request by DelusionalOptimist**
Signed-off-by: Rudraksh Pareek <rudraksh@accuknox.com>

**Purpose of PR?**:

* Adds a new agent for reporting entire state of protected nodes
* Refactors feeder
* Adds policy and log services

**Checklist:**
- [ ] Bug fix. Fixes #<issue number>
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: KubeArmor/Makefile [hunk 1]
```diff
@@ -55,7 +55,7 @@ run: build
 .PHONY: run-container
 run-container: build
 	cd $(CURDIR); sudo rm -f /tmp/kubearmor.log
-	cd $(CURDIR); sudo -E ./kubearmor -logPath=/tmp/kubearmor.log -enableKubeArmorHostPolicy -enableKubeArmorPolicy -k8s=false
+	cd $(CURDIR); sudo -E ./kubearmor -logPath=/tmp/kubearmor.log -enableKubeArmorHostPolicy -enableKubeArmorPolicy -k8s=false -criSocket=unix:///var/run/docker.sock
```
Comment (nyrahul) [inline KubeArmor/Makefile]: Won't this cause issues if non docker container runtime is used by the developer?

File: KubeArmor/common/common.go [hunk 2]
```diff
@@ -375,6 +379,12 @@ func IsK8sEnv() bool {
 	return false
 }
 
+// IsECSEnv Function
+func IsECSEnv() bool {
```
Comment (nyrahul) [inline KubeArmor/common/common.go]: Lets remove ECS specific code.

**Review by nyrahul — DISMISSED** [review]

File: KubeArmor/Makefile (hunk 1 above)
Comment (daemon1024) [reply KubeArmor/Makefile]: We don't have support for that today.

File: KubeArmor/common/common.go (hunk 2 above)
Comment (DelusionalOptimist) [reply KubeArmor/common/common.go]: Oof forgot this. Thanks.

File: KubeArmor/Makefile (hunk 1 above)
Comment (DelusionalOptimist) [reply KubeArmor/Makefile]: Yeah. Unorchestrated non-docker containers has some problems - https://github.com/kubearmor/KubeArmor/issues/1426
Also, without this flag KubeArmor today connects randomly with the path it stats successfully first.
Relevant code - https://github.com/kubearmor/KubeArmor/blob/06ebcedd6e95e87b624dd0d09c87c8114c221bd9/KubeArmor/common/common.go#L398
An unclean way I can think of handling this is having docker as an exception - If containerd socket exists, also check for docker socket?

File: KubeArmor/core/dockerHandler.go [hunk 3]
```diff
@@ -136,6 +147,59 @@ func (dh *DockerHandler) GetContainerInfo(containerID string) (tp.Container, err
...
+		if inspect.HostConfig.NetworkMode.IsNone() || inspect.HostConfig.NetworkMode.IsContainer() {
+			podIP = ""
+		} else if inspect.HostConfig.NetworkMode.IsHost() {
+			podIP = dh.NodeIP
+		} else if inspect.HostConfig.NetworkMode.IsDefault() {
+			podIP = inspect.NetworkSettings.Networks["bridge"].IPAddress
+		} else {
+			networkName := inspect.HostConfig.NetworkMode.NetworkName()
+			podIP = inspect.NetworkSettings.Networks[networkName].IPAddress
+		}
```
Comment (Ankurk99) [inline KubeArmor/core/dockerHandler.go:174]: suggestion: switch case would have been more readable

File: KubeArmor/core/dockerHandler.go [hunk 4]
```diff
@@ -136,6 +147,59 @@ func (dh *DockerHandler) GetContainerInfo(containerID string) (tp.Container, err
...
+		if err == nil {
+			container.LastUpdatedAt = lastUpdatedAt.UTC().String()
+		}
+		// finished at is IsZero until a container exits
+		timeFinished, err := time.Parse(time.RFC3339Nano, inspect.State.FinishedAt)
+		if err == nil && !timeFinished.IsZero() && timeFinished.After(lastUpdatedAt) {
+			lastUpdatedAt = timeFinished
+		}
+
+		// TODO
```
Comment (Ankurk99) [inline KubeArmor/core/dockerHandler.go]: Can you pls briefly explain all the TODOs?

File: KubeArmor/policy/reversePolicyServer.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+			kg.Warnf("Failed to send a HostPolicy status response=[%+v] err=[%s]", res, err.Error())
+			continue
+		}
+	}
+
+	return
+}
+
+// TODO: use single gRPC connection for both the clients
+func (ps *ReversePolicyServer) connectWithRelay() {
```
Comment (Ankurk99) [inline KubeArmor/policy/reversePolicyServer.go]: pls address the go-lints

File: KubeArmor/common/common.go [hunk 6]
```diff
@@ -456,3 +463,61 @@ func WriteToFile(val interface{}, destFile string) error {
...
+		u, repErr := url.ParseRequestURI("http://" + address)
+		if repErr != nil {
+			return "", "", fmt.Errorf("Error while parsing URL: %s", err)
+		}
+
+		addr = u
+	}
+
+	if addr.Port() != "" {
+		host = strings.Split(addr.Host, ":")[0]
```
Comment (Ankurk99) [inline KubeArmor/common/common.go]: Why didn't we use host = addr.Hostname() here?

**Review by Ankurk99 — DISMISSED** [review]
Some suggestions inline. Also pls remove the commented code where not required.

File: KubeArmor/core/dockerHandler.go (hunk 3 above)
Comment (DelusionalOptimist) [reply KubeArmor/core/dockerHandler.go:174]: @Ankurk99 That might not be possible as these functions return bools.

File: KubeArmor/common/common.go (hunk 6 above)
Comment (DelusionalOptimist) [reply KubeArmor/common/common.go]: Fixed. Thanks for the suggestion.

**Review by daemon1024 — APPROVED** [review]

**Review by Aryan-sharma11 — APPROVED** [review]

**Pull request merged**